core/trino-main/src/main/java/io/trino/execution/scheduler/BinPackingNodeAllocatorService.java

```diff
@@ -0,0 +1,435 @@
+    {
+        if (started.compareAndSet(false, true)) {
+            executor.schedule(() -> {
+                while (!stopped.get()) {
+                    try {
+                        // pending acquires are processed when node is released (semaphore is bumped) and periodically (every couple seconds)
+                        // in case node list in cluster have changed.
+                        processSemaphore.tryAcquire(PROCESS_PENDING_ACQUIRES_DELAY_SECONDS, TimeUnit.SECONDS);
+                        processSemaphore.drainPermits();
+                        processPendingAcquires();
+                    }
+                    catch (InterruptedException e) {
+                        currentThread().interrupt();
+                    }
+                    catch (Exception e) {
+                        // ignore to avoid getting unscheduled
+                        log.warn(e, "Error updating nodes");
+                    }
+                }
+            }, 0, TimeUnit.SECONDS);
+        }
+
+        refreshNodePoolSizes();
+        executor.scheduleWithFixedDelay(this::refreshNodePoolSizes, 1, 5, TimeUnit.SECONDS);
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        stopped.set(true);
+        executor.shutdownNow();
+    }
+
+    private void refreshNodePoolSizes()
+    {
+        ImmutableMap.Builder<String, Long> newNodePoolSizes = ImmutableMap.builder();
+
+        Map<String, Optional<MemoryInfo>> workerMemoryInfo = workerMemoryInfoSupplier.get();
+        for (Map.Entry<String, Optional<MemoryInfo>> entry : workerMemoryInfo.entrySet()) {
+            if (entry.getValue().isEmpty()) {
+                continue;
+            }
+            newNodePoolSizes.put(entry.getKey(), entry.getValue().get().getPool().getMaxBytes());
+        }
+        nodePoolSizes.set(newNodePoolSizes.buildOrThrow());
+    }
+
+    @VisibleForTesting
+    synchronized void processPendingAcquires()
+    {
+        // synchronized only for sake manual triggering in test code. In production code it should only be called by single thread
+        Iterator<PendingAcquire> iterator = pendingAcquires.iterator();
+
+        BinPackingSimulation simulation = new BinPackingSimulation(
+                nodeManager.getActiveNodesSnapshot(),
+                nodePoolSizes.get(),
+                allocatedMemory,
+                scheduleOnCoordinator);
+
+        while (iterator.hasNext()) {
+            PendingAcquire pendingAcquire = iterator.next();
+
+            if (pendingAcquire.getFuture().isCancelled()) {
+                // request aborted
+                iterator.remove();
+                continue;
+            }
+
+            BinPackingSimulation.ReserveResult result = simulation.tryReserve(pendingAcquire);
+            switch (result.getStatus()) {
+                case RESERVED:
+                    InternalNode reservedNode = result.getNode().orElseThrow();
+                    pendingAcquire.getFuture().set(reservedNode);
```

Comment:
I feel it's more natural to put this line into the if block. Although it doesn't really make a difference.

Reply:
actually it cannot be in `if` block. Then we would have a race and future could still get cancelled just before I call the `pendingAcquire.getFuture().set(reservedNode);` on it. I need to check for callation after trying to set the value.

Reply:
I think a race exists regardless. Future can be cancelled after you checked cancellation and before you call `updateAllocatedMemory`, right?